src/module.c

```diff
@@ -10846,10 +10998,407 @@ sds genModulesInfoString(sds info) {
+ * callbacks, as well as several flags that modify the behavior of the config. The
+ * supported flags are:
+ * 
+ * * REDISMODULE_CONFIG_DEFAULT: This config can be modified after startup.
+ * * REDISMODULE_CONFIG_IMMUTABLE: This config can only be provided loading time.
+ * * REDISMODULE_CONFIG_SENSITIVE: The value stored in this config is redacted from all logging.
+ * * REDISMODULE_CONFIG_HIDDEN: The name is hidden from `CONFIG GET` with pattern matching. 
+ * * REDISMODULE_CONFIG_PROTECTED: This config will be only be modifiable based off the servers protected config value.
+ * * REDISMODULE_CONFIG_DENY_LOADING: This config is not modifiable while the server is loading data.
+ * * REDISMODULE_CONFIG_MEMORY: For numeric configs, this config will convert data unit notations into their byte equivalent.
+ *
+ * Default values are used explicitly to compare to on a config rewrite.
+ *
+ * Note on strings that the previous string and the new string will both be freed and a set callback should always retain one
+ * 
+ * Example implementation:
+ * 
+ *     RedisModuleString *strval;
+ *     int adjustable = 1;
+ *     RedisModuleString *getStringConfigCommand(const char *name, void *privdata) {
+ *         return strval;
+ *     }
+ *     
+ *     int setStringConfigCommand(const char *name, RedisModuleString *new, void *privdata, const char **err) {
+          if (adjustable) {
+              RedisModule_RetainString(NULL, new);
+              strval = new;
+              return REDISMODULE_OK;
+          }
+ *        if (strval) RedisModule_RetainString(NULL, strval);
+ *        return REDISMODULE_ERR;
+ *     }
+ *     ...
+ *     RedisModule_RegisterStringConfig(ctx, "string", NULL, REDISMODULE_CONFIG_DEFAULT, getStringConfigCommand, setStringConfigCommand, NULL, NULL);
+ * 
+ * If the registration fails, REDISMODULE_ERR is returned and one of the following 
+ * errno is set:
+ * * EINVAL: The provided flags are invalid for the registration.
+ * * EALREADY: The provided configuration name is already used. */
+int RM_RegisterStringConfig(RedisModuleCtx *ctx, const char *name, RedisModuleString *default_val, unsigned int flags, RedisModuleConfigGetStringFunc getfn, RedisModuleConfigSetStringFunc setfn, RedisModuleConfigApplyFunc applyfn, void *privdata) {
+    RedisModule *module = ctx->module;
+    if (moduleConfigValidityCheck(module, name, flags, REDISMODULE_CONFIG_STRING)) {
+        return REDISMODULE_ERR;
+    }
+    ModuleConfig newStringConfig = createModuleConfig(sdsnew(name), REDISMODULE_CONFIG_STRING, applyfn, privdata, module);
+    addDefaultAndCallbacksToConfig(newStringConfig, getfn, setfn, string);
+    addModuleConfigtoList(module, &newStringConfig);
+    flags = maskModuleConfigFlags(flags);
+    addModuleStringConfig(module->name, name, flags, module->module_configs->tail->value, default_val ? default_val->ptr : NULL);
```

Comment:
If you take my earlier suggestion to update `createModuleConfig`
```suggestion
    ModuleConfig *newStringConfig = createModuleConfig(sdsnew(name), REDISMODULE_CONFIG_STRING, applyfn, privdata, module);
    addDefaultAndCallbacksToConfig(newStringConfig, getfn, setfn, string);
    addModuleConfigtoList(module, newStringConfig);
    flags = maskModuleConfigFlags(flags);
    addModuleStringConfig(module->name, name, flags, newStringConfig, default_val ? default_val->ptr : NULL);
```

Reply:
Also, we should duplicate the default value and decrRefCount on the RedisModule string, we only need to keep around the sds value.